src/newTest/java/org/opensearch/test/framework/cluster/LocalOpenSearchCluster.java

```diff
@@ -0,0 +1,610 @@
+                }
+            } else if (!dataNodes.isEmpty()) {
+                for (NodeInfo nodeInfo : dataNodes) {
+                    if (hasHttpAddressPredicate.test(nodeInfo)) {
+                        final TransportAddress his = nodeInfo.getInfo(HttpInfo.class).address().publishAddress();
+                        clusterInfo.httpPort = his.getPort();
+                        clusterInfo.httpHost = his.getAddress();
+                        break;
+                    }
+                }
+            } else {
+                for (NodeInfo nodeInfo : nodes) {
+                    if (hasHttpAddressPredicate.test(nodeInfo)) {
+                        final TransportAddress his = nodeInfo.getInfo(HttpInfo.class).address().publishAddress();
+                        clusterInfo.httpPort = his.getPort();
+                        clusterInfo.httpHost = his.getAddress();
+                        break;
+                    }
+                }
+            }
+
+            for (NodeInfo nodeInfo : nodes) {
+                clusterInfo.httpAdresses.add(nodeInfo.getInfo(HttpInfo.class).address().publishAddress());
+            }
+        } catch (final OpenSearchTimeoutException e) {
+            throw new IOException("timeout, cluster does not respond to health request, cowardly refusing to continue with operations");
+        }
+        return clusterInfo;
+    }
+
+    @Override
+    public String toString() {
+        return "\nES Cluster " + clusterName + "\nmaster nodes: " + masterNodes + "\n  data nodes: " + dataNodes + "\nclient nodes: " + clientNodes
+                + "\n";
+    }
+
+    private static List<String> toHostList(Collection<Integer> ports) {
+        return ports.stream().map(port -> "127.0.0.1:" + port).collect(Collectors.toList());
+    }
+
+    private String createNextNodeName(NodeSettings nodeSettings) {
+        List<Node> nodes;
+        String nodeType;
+
+        if (nodeSettings.masterNode) {
+            nodes = this.masterNodes;
+            nodeType = "master";
+        } else if (nodeSettings.dataNode) {
+            nodes = this.dataNodes;
+            nodeType = "data";
+        } else {
+            nodes = this.clientNodes;
+            nodeType = "client";
+        }
+
+        return nodeType + "_" + nodes.size();
+    }
+
+    public class Node implements OpenSearchClientProvider {
+        private final String nodeName;
+        private final NodeSettings nodeSettings;
+        private final File nodeHomeDir;
+        private final File dataDir;
+        private final File logsDir;
+        private final int transportPort;
+        private final int httpPort;
+        private final InetSocketAddress httpAddress;
+        private final InetSocketAddress transportAddress;
+        private PluginAwareNode node;
+        private boolean running = false;
+        private boolean portCollision = false;
+
+        Node(NodeSettings nodeSettings, int transportPort, int httpPort) {
```

Comment:
Maybe rename to `RunningNode`?

Reply:
What would be the rationale for that?

Reply:
There is a state where a node is being configured to start, when it has starting and when it is actively running.  The node case is managing its lifecycle in this single class definition, it might be easier to manage these aspects separately so node/cluster doesn't need to check the state of the nodes once it has reached a threshold of availability.